tests/unit/test_actions.py

```diff
@@ -0,0 +1,129 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Jenkins-k8s charm actions unit tests."""
+
+import secrets
+
+# Need access to protected functions for testing
+import typing
+from unittest.mock import MagicMock
+
+import ops
+import pytest
+
+import charm
+import jenkins
+from charm import JenkinsK8sOperatorCharm
+
+from .types_ import HarnessWithContainer
+
+
+def test_on_get_admin_password_action_container_not_ready(
+    harness_container: HarnessWithContainer,
+):
+    """
+    arrange: given a jenkins container that is not connectable.
+    act: when get-admin-password action is run.
+    assert: the event is failed.
+    """
+    harness_container.harness.set_can_connect(
+        harness_container.harness.model.unit.containers["jenkins"], False
+    )
+    mock_event = MagicMock(spec=ops.ActionEvent)
```

Comment:
@benhoyt it could be interesting to make this easier, like providing a fake event class or something like that for use in tests?

Reply:
It is relatively minor, it would be handy for the framework to set the `spec` argument rather than it being done by users of the framework so that we are sure that we're using the right `spec` class

Reply:
If I understand the test correctly we're just testing an action here (and that it called `fail`). The new support for testing actions with `Harness.run_action` should be good for this. See https://github.com/canonical/operator/pull/1053

@tonyandrewmeyer added this recently. It's merged in the `main` branch, and will be included in the next `ops` release at the end of November, 2.9.0.